lib/sycamore/sycamore/writers/pinecone_writer.py

```diff
@@ -0,0 +1,175 @@
+        self._batch_size = batch_size
+        self._api_key = api_key
+
+    def on_write_start(self) -> None:
+        from pinecone import Pinecone
+        from pinecone.core.client.exceptions import NotFoundException
+
+        pc = Pinecone(api_key=self._api_key)
+        try:
+            pc.describe_index(self._index_name)
+        except NotFoundException:
+            if self._dimensions is None:
+                raise ValueError(f"dimensions must be specified in order to create a new index {self._index_name}")
+            if self._index_spec is None:
+                raise ValueError(f"index spec must be supplied in order to create a new index {self._index_name}")
+            pc.create_index(
+                name=self._index_name,
+                dimension=self._dimensions,
+                spec=self._index_spec,
+                metric=self._distance_metric,
+                timeout=None,
+            )
+
+    def write(self, blocks: Iterable[Block], ctx: TaskContext) -> Any:
+        from pinecone.grpc import PineconeGRPC
+
+        builder = DelegatingBlockBuilder()
+        for block in blocks:
+            builder.add_block(block)
+        block = builder.build()
+
+        pc = PineconeGRPC(api_key=self._api_key)
+        index = pc.Index(self._index_name)
+
+        objects = self._extract_pinecone_objects(block)
+        obj_it = iter(objects)
+        batch = list(itertools.islice(obj_it, self._batch_size))
+        async_results = []
+        while len(batch) > 0:
+            async_results.append(index.upsert(vectors=batch, namespace=self._namespace, async_req=True))
+            batch = list(itertools.islice(obj_it, self._batch_size))
+        for r in async_results:
+            r.result()
+
+    @staticmethod
+    def _extract_pinecone_objects(block: Block) -> Iterable[Vector]:
+        def _flatten_metadata(data: Union[dict, list, tuple], prefix="") -> Iterable[Tuple[Any, Any]]:
+            iterator = []  # type: ignore
+            if isinstance(data, dict):
+                iterator = data.items()  # type: ignore
+            if isinstance(data, (list, tuple)):
+                iterator = enumerate(data)  # type: ignore
+            items = []
+            for k, v in iterator:
+                if isinstance(v, (dict, list, tuple)):
+                    if isinstance(v, (list, tuple)) and all(isinstance(innerv, str) for innerv in v):
+                        items.append(((str(k) if len(prefix) == 0 else f"{prefix}.{k}"), v))
+                    else:
+                        inner_values = _flatten_metadata(v, prefix=(str(k) if len(prefix) == 0 else f"{prefix}.{k}"))
+                        items.extend([(innerk, innerv) for innerk, innerv in inner_values])
+                elif v is not None:
+                    items.append(((str(k) if len(prefix) == 0 else f"{prefix}.{k}"), v))
+            return items
+
+        def _metadata_special_cases(data: dict) -> dict:
```

Comment:
Is there a way to generalize this, so we don't have to list every field? Maybe handle things by type? Are there ever cases where we are going to want to handle two properties/fields of the same type differently? 

I also wonder if the Table object in table elements will cause trouble: https://github.com/aryn-ai/sycamore/blob/3914113955d87308f6f4c04afa9a59618357d401/lib/sycamore/sycamore/data/element.py#L121 ?

Reply:
I'd suggest adding a .to_simple_type() method to the metadata special cases; then the logic for conversion goes becomes part of the type rather than part of the writer.  Unclear what to do in cases where it isn't a simple type, probably a warning and skip the field is the best you can do. 

The other two writers are currently just taking whatever is in properties and passing it through. I have no idea what that will do with complex types.

Reply:
moving around the logic to better match how the weaviate and opensearch connectors deal with Document -> DB record lets me drop this fn entirely. Also changing it to actively select what Document properties to load rather than just flattening the entire document means I can fully drop the table (as the other writers also do).

Now, we might want to write tables to our other sources as well, but that's a separate issue.